### PR TITLE
Clarify `wait(::Task)` documentation

### PR DESCRIPTION
**Pull request by Seelengrab**
Small follow-up to https://github.com/JuliaLang/julia/pull/53685, since having critical documentation for a specific method be only mentioned in a catch-all docstring isn't quite as helpful as it could be (e.g. when looking at inline documentation with LSP, where it happens to know that the variable being passed is a `Task`). In addition, this also documents that `wait` on a `Task` will throw if given `current_task`.

I've also split the function into a prototype and the actual implementation, so that the same "more specific docs at the method definition" transform can be applied for `GenericCondition` too.

### CODE REVIEW

**Comment by Seelengrab** [comment]
Failure is real, it's due to the change in exception type - https://buildkite.com/julialang/julia-master/builds/34661#018e3802-f274-4321-9018-f01d228c041d/916-1696

The question is - is this an alright change, or considered breaking? After all, it was undocumented that this threw at all...

File: base/task.jl [hunk 1]
```diff
@@ -356,8 +356,18 @@ function _wait2(t::Task, waiter::Task)
...
+Wait for a `Task` to finish.
+
+The keyword `throw` (defaults to `true`) controls whether a failed task results
+in an error, thrown as a [`TaskFailedException`](@ref) which wraps the failed task.
+
+Throws an error if `t` is the currently running task, to prevent deadlocks.
+"""
 function wait(t::Task; throw=true)
-    t === current_task() && error("deadlock detected: cannot wait on current task")
+    t === current_task() && throw(ArgumentError("deadlock detected: cannot wait on current task"))
```
[the remark this answers is not in the document]
Comment (vtjnash) [reply base/task.jl:370]: ConcurrencyViolationError seems reasonable here. It is just what its name says: any detected errors with concurrency. It does not seem to be an ArgumentError.

**Comment by Seelengrab** [comment]
@vtjnash I'd like to add a docstring to `ConcurrencyViolationError` in a different PR, but noticed that pretty much nothing in `boot.jl` (where this is defined) has a docstring. Are docstrings for things defined in there added somewhere else due to bootstrapping issues with inline docstrings, or should I add it in there?

**Comment by vtjnash** [comment]
base/docs/basedocs.jl

**Comment by Seelengrab** [comment]
Missed that there's a conflict between `Core.throw` and the `throw` keyword, should be good now. The question is though if the keyword should be renamed because of this conflict :thinking:

**Comment by Seelengrab** [comment]
Is this good to go or should the keyword be renamed after all?

**Pull request merged**